docs/BEST_PRACTICES.md

```diff
@@ -143,3 +143,53 @@ install_requires=[
     "singer-python"
 ]
 ```
+
+## Parent-Child Relationships
+
+One of the most challenging API structures for ELT, common with REST APIs,
+is the existence of a parent-child relationship. For example, resources
+can form a tree like a structure of `organization -> project -> ticket ->
+comment`, where each sub-object requires the ID of the parent object. A
+request for `tickets` in this structure may look something like `GET
+/organizations/{org_id}/projects/{project_id}/tickets`.
+
+This is an efficient design for applications, but can pose challenges for
+bulk extraction. The specific challenges are very specific to an
+individual API's underlying implementation, but there are some best
+practices that can provide guidance in many cases.
+
+1. **Reduce Stream Dependencies** - Keep logic for child streams separate
+   from parent streams. Each stream should be able to be extracted without
+   requiring the parent to be selected and without complex boolean logic
+   to determine what data to emit.
+   - This opens up options in implementation that can provide efficiencies
+     like only requesting the parent object's ID field for the child
+     object's sync, and generally results in more readable code.
+   - For REST APIs that provide a client library, the library functions
+     usually try to make efficient use of this pattern by lazy loading the
+     parent object and only making HTTP requests as necessary.
+   - [Example
+   tap-shopify](https://github.com/singer-io/tap-shopify/blob/v1.2.9/tap_shopify/streams/order_refunds.py#L28-L31)
+2. **Signpost State Keeping** - Set a maximum bookmark value before the
+   sync begins and store bookmarks only up to that value. For large data
+   sets, records are being updated as the tap is requesting all of the
+   data. As the tap's extraction intersects with the data changing in real
+   time, records can be missed, but setting a "signpost" bookmark as a
+   maximum value to be stored can help with many race conditions and
+   sorting issues.
```

Comment:
@dmosorast - Thanks for creating this content! I found the second sentence in this section a little hard to follow personally, specifically why/how records get missed. I have been struggling to document something similar myself so I took a stab below. It's hard to explain concisely but I think we're getting close.

```suggestion
   data. As the tap's extraction reads records that may be changing in real
   time, setting a "signpost" bookmark as a
   maximum value to be stored can help with many race conditions and
   sorting issues. This signpost prevents the bookmark from advancing to
   a point where subsequent executions of the tap would miss records modified
   within the batch interval.
```

Reply:
Yup, this does read better. Thanks for the suggestion!